File: scala/core/src/main/scala/blockchain/consensus/BlockHeights.scala
package blockchain.consensus

import blockchain.{BlockIdTree, BlockSourcedState}
import blockchain.models.*
import blockchain.Store
import cats.effect.{Async, Resource}
import cats.implicits.*

object BlockHeights:
  type State[F[_]] = Long => F[Option[BlockId]]
  type BSS[F[_]] = BlockSourcedState[F, State[F]]

  def make[F[_]: Async](
      store: Store[F, Long, BlockId],
      initialEventId: F[BlockId],
      blockTree: BlockIdTree[F],
      currentEventChanged: BlockId => F[Unit],
      fetchHeader: BlockId => F[BlockHeader]
  ): Resource[F, BSS[F]] =
    BlockSourcedState.make[F, State[F]](
      Async[F].delay(store.get),
      initialEventId = initialEventId,
      (state, id) => fetchHeader(id).map(_.height).flatTap(store.put(_, id)).as(state),
      (state, id) => fetchHeader(id).map(_.height).flatTap(store.remove).as(state),
      blockTree,
      currentEventChanged
    )

File: dart/protocol/lib/src/blockchain/consensus/block_heights.dart
import 'package:fixnum/fixnum.dart';
import 'package:giraffe_protocol/protocol.dart';
import 'package:giraffe_protocol/src/blockchain/block_id_tree.dart';
import 'package:giraffe_protocol/src/blockchain/block_sourced_state.dart';
import 'package:giraffe_sdk/sdk.dart';

import '../store.dart';

class BlockHeights {
  static BlockHeightsBSS make(
      Store<Int64, BlockId> store,
      BlockId currentId,
      BlockIdTree blockIdTree,
      Future<void> Function(BlockId) blockIdChanged,
      FetchHeader fetchHeader) {
    return BlockSourcedStateImpl(
      applyBlock: (state, id) async =>
          store.put((await fetchHeader(id)).height, id).then((_) => state),
      unapplyBlock: (state, id) async =>
          store.remove((await fetchHeader(id)).height).then((_) => state),
      blockIdTree: blockIdTree,
      state: store.get,
      currentId: currentId,
      blockIdChanged: blockIdChanged,
    );
  }
}

typedef BlockHeightsState = Future<BlockId?> Function(Int64);
typedef BlockHeightsBSS = BlockSourcedState<BlockHeightsState>;
